diff --git a/fs/btrfs/compression.h b/fs/btrfs/compression.h
index 4691a84ca838..8399e618eaeb 100644
--- a/fs/btrfs/compression.h
+++ b/fs/btrfs/compression.h
@@ -39,6 +39,7 @@ static_assert((BTRFS_MAX_COMPRESSED % PAGE_SIZE) == 0);
 #define BTRFS_MAX_UNCOMPRESSED		(SZ_128K)
 
 #define	BTRFS_ZLIB_DEFAULT_LEVEL		3
+#define	BTRFS_ZSTD_DEFAULT_LEVEL		3
 
 struct compressed_bio {
 	/* Number of compressed pages in the array */
diff --git a/fs/btrfs/defrag.c b/fs/btrfs/defrag.c
index f015fa1b6301..4579524fdea8 100644
--- a/fs/btrfs/defrag.c
+++ b/fs/btrfs/defrag.c
@@ -1371,7 +1371,7 @@ int btrfs_defrag_file(struct inode *inode, struct file_ra_state *ra,
 	u64 last_byte;
 	bool do_compress = (range->flags & BTRFS_DEFRAG_RANGE_COMPRESS);
 	bool ra_allocated = false;
-	int compress_type = BTRFS_COMPRESS_ZLIB;
+	int compress_type = BTRFS_COMPRESS_ZSTD;
 	int ret = 0;
 	u32 extent_thresh = range->extent_thresh;
 	pgoff_t start_index;
diff --git a/fs/btrfs/disk-io.c b/fs/btrfs/disk-io.c
index 3df5477d48a8..48caaf757cd2 100644
--- a/fs/btrfs/disk-io.c
+++ b/fs/btrfs/disk-io.c
@@ -2824,7 +2824,7 @@ void btrfs_init_fs_info(struct btrfs_fs_info *fs_info)
 	fs_info->stripesize = 4096;
 
 	/* Default compress algorithm when user does -o compress */
-	fs_info->compress_type = BTRFS_COMPRESS_ZLIB;
+	fs_info->compress_type = BTRFS_COMPRESS_ZSTD;
 
 	fs_info->max_extent_size = BTRFS_MAX_EXTENT_SIZE;
 
diff --git a/fs/btrfs/ioctl.c b/fs/btrfs/ioctl.c
index 55f3ba6a831c..84964d35faf1 100644
--- a/fs/btrfs/ioctl.c
+++ b/fs/btrfs/ioctl.c
@@ -361,7 +361,7 @@ int btrfs_fileattr_set(struct mnt_idmap *idmap,
 
 		comp = btrfs_compress_type2str(fs_info->compress_type);
 		if (!comp || comp[0] == 0)
-			comp = btrfs_compress_type2str(BTRFS_COMPRESS_ZLIB);
+			comp = btrfs_compress_type2str(BTRFS_COMPRESS_ZSTD);
 	} else {
 		binode_flags &= ~(BTRFS_INODE_COMPRESS | BTRFS_INODE_NOCOMPRESS);
 	}
diff --git a/fs/btrfs/super.c b/fs/btrfs/super.c
index 7e44ccaf348f..2fcd16860e20 100644
--- a/fs/btrfs/super.c
+++ b/fs/btrfs/super.c
@@ -333,8 +333,8 @@ static int btrfs_parse_param(struct fs_context *fc, struct fs_parameter *param)
 	case Opt_compress:
 	case Opt_compress_type:
 		if (opt == Opt_compress || opt == Opt_compress_force) {
-			ctx->compress_type = BTRFS_COMPRESS_ZLIB;
-			ctx->compress_level = BTRFS_ZLIB_DEFAULT_LEVEL;
+			ctx->compress_type = BTRFS_COMPRESS_ZSTD;
+			ctx->compress_level = BTRFS_ZSTD_DEFAULT_LEVEL;
 			btrfs_set_opt(ctx->mount_opt, COMPRESS);
 			btrfs_clear_opt(ctx->mount_opt, NODATACOW);
 			btrfs_clear_opt(ctx->mount_opt, NODATASUM);
