diff --git a/man/man.md b/man/man.md
index 74f83d1..decea7f 100644
--- a/man/man.md
+++ b/man/man.md
@@ -132,11 +132,11 @@ F3
 
 +
 
-:   Increase interval by .5 seconds.
+:   Increase interval by .1 seconds.
 
 -
 
-:   Decrease interval by .5 seconds (As long as it's positive).
+:   Decrease interval by .1 seconds (As long as it's positive).
 
 Tab
 
diff --git a/src/app.rs b/src/app.rs
index 386915a..6782f0e 100644
--- a/src/app.rs
+++ b/src/app.rs
@@ -470,14 +470,14 @@ impl<'a> App<'a> {
     ///
     fn increase_interval(&mut self) {
         let cur_interval = *self.interval.read().unwrap();
-        self.set_interval(cur_interval + 0.5);
+        self.set_interval(cur_interval + 0.1);
     }
 
     ///
     fn decrease_interval(&mut self) {
         let cur_interval = *self.interval.read().unwrap();
-        if cur_interval > 0.5 {
-            self.set_interval(cur_interval - 0.5);
+        if cur_interval >= 0.1 {
+            self.set_interval(cur_interval - 0.1);
         }
     }
 
diff --git a/src/help.rs b/src/help.rs
index 63d1e28..8a94116 100644
--- a/src/help.rs
+++ b/src/help.rs
@@ -90,8 +90,8 @@ fn gen_help_text<'a>() -> Vec<Spans<'a>> {
         Spans::from(" - [F2] key  ... change output mode as stderr."),
         Spans::from(" - [F3] key  ... change output mode as output(stdout/stderr set.)"),
         // change interval
-        Spans::from(" - [+] key ... Increase interval by .5 seconds."),
-        Spans::from(" - [-] key ... Decrease interval by .5 seconds."),
+        Spans::from(" - [+] key ... Increase interval by .1 seconds."),
+        Spans::from(" - [-] key ... Decrease interval by .1 seconds."),
         // change use area
         Spans::from(" - [Tab] key ... toggle current area at history or watch."),
         // filter text inpu
