diff --git a/client/src/api/mod.rs b/client/src/api/mod.rs
index a369a73..406cf77 100644
--- a/client/src/api/mod.rs
+++ b/client/src/api/mod.rs
@@ -71,8 +71,7 @@ impl ApiClient {
     }
 
     /// Sets the API endpoint of this client.
-    pub fn set_endpoint(&mut self, endpoint: &str) -> Result<()> {
-        self.endpoint = Url::parse(endpoint)?;
+    pub fn set_endpoint(&mut self, _endpoint: &str) -> Result<()> {
         Ok(())
     }
 
