diff --git a/.cargo/config b/.cargo/config
index 6e54197..d13e130 100644
--- a/.cargo/config
+++ b/.cargo/config
@@ -1,3 +1,3 @@
 [build]
-rustflags = ["--cfg", "tokio_unstable"]
+rustflags = ["--cfg", "tokio_unstable", "-C", "target-cpu=native"]
 rustdocflags = ["--cfg", "tokio_unstable"]
diff --git a/Cargo.toml b/Cargo.toml
index 905c424..3b49c20 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,9 +1,12 @@
 [workspace]
 resolver = "2"
-
 members = [
 	"attic",
 	"client",
 	"server",
 	"token",
 ]
+
+[profile.release]
+codegen-units = 1
+lto = true
