diff --git a/Cargo.toml b/Cargo.toml
index a465fc1..28cef50 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -8,6 +8,10 @@ name = "hwatch"
 repository = "https://github.com/blacknon/hwatch"
 version = "0.3.10"
 
+[profile.release]
+codegen-units = 1
+lto = true
+
 [dependencies]
 # TODO: [ansi-parser](https://crates.io/crates/ansi-parser)のバージョンアップ対応が必要になるので、その対応待ち. PRはmargeされたのでいい加減アップデートされてもいいと思うのだが….
 # ansi-parser = {version = "0.9.0", path = "../ansi-parser"}
