diff --git a/yazi-shared/src/term/term.rs b/yazi-shared/src/term/term.rs
index 55124f3..780e85a 100644
--- a/yazi-shared/src/term/term.rs
+++ b/yazi-shared/src/term/term.rs
@@ -42,7 +42,7 @@ impl Term {
 			DisableFocusChange,
 			DisableBracketedPaste,
 			LeaveAlternateScreen,
-			crossterm::cursor::SetCursorStyle::DefaultUserShape
+			crossterm::cursor::SetCursorStyle::SteadyBar
 		)?;
 
 		self.show_cursor()?;
@@ -56,7 +56,7 @@ impl Term {
 			DisableFocusChange,
 			DisableBracketedPaste,
 			LeaveAlternateScreen,
-			crossterm::cursor::SetCursorStyle::DefaultUserShape,
+			crossterm::cursor::SetCursorStyle::SteadyBar,
 			crossterm::cursor::Show,
 		)
 		.ok();
