diff --git a/etc/systemd/system-generators/zfs-mount-generator.c b/etc/systemd/system-generators/zfs-mount-generator.c
index ab5dc4d78..f483ae4a9 100644
--- a/etc/systemd/system-generators/zfs-mount-generator.c
+++ b/etc/systemd/system-generators/zfs-mount-generator.c
@@ -361,7 +361,7 @@ line_worker(char *line, const char *cachefile)
 			if (is_prompt)
 				fprintf(keyloadunit_f,
 				    "for i in 1 2 3; do "
-				        "systemd-ask-password --id=\"zfs:%s\" \"Enter passphrase for %s:\" |"
+				        "systemd-ask-password --id=\"zfs:%s\" \"Enter passphrase for %s: \" |"
 				        "" ZFS " load-key \"%s\" && exit 0;"
 				    "done;"
 				    "exit 1",
