diff --git a/main.c b/main.c
index 51789cd..c2779df 100644
--- a/main.c
+++ b/main.c
@@ -129,6 +129,9 @@ static void handle_active_selection_motion(struct slurp_seat *seat, struct slurp
 static void pointer_handle_enter(void *data, struct wl_pointer *wl_pointer,
 		uint32_t serial, struct wl_surface *surface,
 		wl_fixed_t surface_x, wl_fixed_t surface_y) {
+	if (surface_x < 0 || surface_y < 0)
+		return;
+
 	struct slurp_seat *seat = data;
 	struct slurp_output *output = output_from_surface(seat->state, surface);
 	if (output == NULL) {
