diff --git a/CMakeLists.txt b/CMakeLists.txt
index ef077d8b2c9814217615e00ec8f576bb95dd5fb1..dd29616e47c78caaf2cbecc56ad976cdef523c87 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -304,11 +304,6 @@ install(CODE "execute_process( \
 install(FILES ${CMAKE_SOURCE_DIR}/example/hyprland.desktop
         DESTINATION ${CMAKE_INSTALL_DATADIR}/wayland-sessions)
 
-# wallpapers
-file(GLOB_RECURSE WALLPAPERS "assets/wall*")
-install(FILES ${WALLPAPERS}
-        DESTINATION ${CMAKE_INSTALL_DATADIR}/hyprland)
-
 # default config
 install(FILES ${CMAKE_SOURCE_DIR}/example/hyprland.conf
         DESTINATION ${CMAKE_INSTALL_DATADIR}/hyprland)
diff --git a/assets/meson.build b/assets/meson.build
index 8c4a60ec6f0e5e7e5ecbe847d7ecc429f12dec97..befebe86a7196bf62c233ecc6201bbfabccbd535 100644
--- a/assets/meson.build
+++ b/assets/meson.build
@@ -1,7 +1 @@
-wallpapers = ['0', '1', '2']
-
-foreach type : wallpapers
-  install_data(f'wall@type@.png', install_dir: join_paths(get_option('datadir'), 'hyprland'), install_tag: 'runtime')
-endforeach
-
 install_data('hyprland-portals.conf', install_dir: join_paths(get_option('datadir'), 'xdg-desktop-portal'), install_tag: 'runtime')
diff --git a/assets/wall0.png b/assets/wall0.png
deleted file mode 100644
index 73bdeef833bda8bc998e35302e36eccc7e7fc418..0000000000000000000000000000000000000000
Binary files a/assets/wall0.png and /dev/null differ
diff --git a/assets/wall1.png b/assets/wall1.png
deleted file mode 100644
index 53e4c90d9b40fd027d635af9a72222a03cbd2c0d..0000000000000000000000000000000000000000
Binary files a/assets/wall1.png and /dev/null differ
diff --git a/assets/wall2.png b/assets/wall2.png
deleted file mode 100644
index 9ade47207f731e74e109e2b0babdedbc80cb7608..0000000000000000000000000000000000000000
Binary files a/assets/wall2.png and /dev/null differ
